tools/slabratetop.py

```diff
@@ -65,6 +68,81 @@
 // 5.9, but it does not hurt to have it here for versions 5.4 to 5.8.
 struct memcg_cache_params {};
 
+// introduce kernel interval slab structure and slab_address() function, solved
+// 'undefined' error for >=5.16. TODO: we should fix this workaround if BCC
+// framework support BTF/CO-RE.
+struct slab {
+        unsigned long __page_flags;
+
+#if defined(CONFIG_SLAB)
+
+        struct kmem_cache *slab_cache;
+        union {
+                struct {
+                        struct list_head slab_list;
+                        void *freelist; /* array of free object indexes */
+                        void *s_mem;    /* first object */
+                };
+                struct rcu_head rcu_head;
+        };
+        unsigned int active;
+
+#elif defined(CONFIG_SLUB)
+
+        struct kmem_cache *slab_cache;
+        union {
+                struct {
+                        union {
+                                struct list_head slab_list;
+#ifdef CONFIG_SLUB_CPU_PARTIAL
+                                struct {
+                                        struct slab *next;
+                                        int slabs;      /* Nr of slabs left */
+                                };
+#endif
+                        };
+                        /* Double-word boundary */
+                        void *freelist;         /* first free object */
+                        union {
+                                unsigned long counters;
+                                struct {
+                                        unsigned inuse:16;
+                                        unsigned objects:15;
+                                        unsigned frozen:1;
+                                };
+                        };
+                };
+                struct rcu_head rcu_head;
+        };
+        unsigned int __unused;
+
+#elif defined(CONFIG_SLOB)
+
+        struct list_head slab_list;
+        void *__unused_1;
+        void *freelist;         /* first free block */
+        long units;
+        unsigned int __unused_2;
+
+#else
+#error "Unexpected slab allocator configured"
+#endif
+
+        atomic_t __page_refcount;
+#ifdef CONFIG_MEMCG
+        unsigned long memcg_data;
+#endif
+};
+
+#define slab_folio(s)           (_Generic((s),                          \
+        const struct slab *:    (const struct folio *)s,                \
+        struct slab *:          (struct folio *)s))
+
+static inline void *slab_address(const struct slab *slab)
+{
+        return folio_address(slab_folio(slab));
```

Comment:
Please fix the following warnings for kernel < 5.16.
```
/virtual/main.c:82:16: warning: implicit declaration of function 'folio_address' is invalid in C99 [-Wimplicit-function-declaration]
        return folio_address(slab_folio(slab));
               ^
/virtual/main.c:82:16: warning: incompatible integer to pointer conversion returning 'int' from a function with result type 'void *' [-Wint-conversion]
        return folio_address(slab_folio(slab));
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
```

Reply:
Hi, Hengqi, Could we just make slab_address() return NULL? because it's never used in slabratetop.py, we don't have to solve folio_address() for different kernel version, do you agree with me?

Reply:
This does solve the problem:

```diff
+static inline void *slab_address(const struct slab *slab)
+{
+    return NULL;
+}
```